kiro-synth-core/src/effects/delay.rs

```diff
@@ -0,0 +1,77 @@
+use crate::float::Float;
+use std::collections::VecDeque;
+
+struct DelayLine<F: Float> {
+  sample_rate: F,
+  buffer: VecDeque<F>,
+}
+
+impl<F: Float> DelayLine<F> {
+  pub fn new(sample_rate: F) -> Self {
+    let delay_max = F::val(2.0);
+    let len = (sample_rate * delay_max).ceil().to_usize().unwrap();
+    Self {
+      sample_rate,
+      buffer: VecDeque::<F>::with_capacity(len),
+    }
+  }
+
+  pub fn update(&mut self, input: F) {
+    self.buffer.pop_front();
+    self.buffer.push_back(input);
+  }
+
+  pub fn get(&self, index: F) -> F {
+    let index = (index * self.sample_rate).to_usize().unwrap();
+    *self.buffer.iter().nth_back(index).unwrap_or(&F::zero())
+  }
+}
+
+pub struct Delay<F: Float> {
+  delay: F,
+  feedback: F,
+  mix: F,
+  delayline: DelayLine<F>,
+}
+
+impl<F: Float> Delay<F> {
+  pub fn new(sample_rate: F) -> Self {
+    Self {
+      delay: F::one(),     // delay in seconds NOTE it must never be 0
+      feedback: F::zero(), // between 0 and 1
+      mix: F::zero(),      // between 0 and 1
+      delayline: DelayLine::<F>::new(sample_rate),
+    }
+  }
+
+  pub fn set_delay(&mut self, delay: F) {
```

Comment:
how do we make sure that it never has a value of 0 ?

Reply:
If we want to check it here, we have 3 options:
- Return a `Result<(), E>`
- `panic!`
- Ignore and don't change

But I guess that the best thing is to put limits in the component wrapping this code (e.g. min and max values for the parameters).